lib/dat-tcp.rb

```diff
@@ -95,103 +98,64 @@ def halt(wait = false)
-      @worker_pool = build_worker_pool
-      add_client_sockets_from_fds client_file_descriptors
-      @worker_pool.start
-      process_inputs while @signal.start?
-      logger.info "Stopping work loop..."
-      shutdown_worker_pool unless @signal.halt?
+    def work_loop(passed_client_fds)
+      setup(passed_client_fds)
+      accept_client_connections while @signal.start?
     rescue StandardError => exception
       logger.error "Exception occurred, stopping server!"
       logger.error "#{exception.class}: #{exception.message}"
       logger.error exception.backtrace.join("\n")
     ensure
-      unless @signal.pause?
-        logger.info "Closing TCP server connection"
-        stop_listen
-      end
-      clear_thread
-      logger.info "Stopped work loop"
+      teardown
     end
 
-    def build_worker_pool
-      wp = DatWorkerPool.new(
-        @min_workers,
-        @max_workers
-      ){ |socket| serve(socket) }
-
-      # add any configured callbacks
-      self.worker_start_procs.each do |cb|
-        wp.on_worker_start(&cb)
-      end
-      self.worker_shutdown_procs.each do |cb|
-        wp.on_worker_shutdown(&cb)
-      end
-      self.worker_sleep_procs.each do |cb|
-        wp.on_worker_sleep(&cb)
-      end
-      self.worker_wakeup_procs.each do |cb|
-        wp.on_worker_wakeup(&cb)
-      end
-
-      wp
-    end
-
-    def add_client_sockets_from_fds(file_descriptors)
-      (file_descriptors || []).each do |file_descriptor|
-        @worker_pool.add_work TCPSocket.for_fd(file_descriptor)
+    def setup(passed_client_fds)
+      logger.info "Starting work loop..."
+      @worker_pool.start
+      (passed_client_fds || []).each do |fd|
+        @worker_pool.push TCPSocket.for_fd(fd)
       end
     end
 
-    def process_inputs
-      ready_inputs, _, _ = IO.select([ @tcp_server, @signal_reader ])
-      accept_connection if ready_inputs.include?(@tcp_server)
-      process_signal    if ready_inputs.include?(@signal_reader)
-    end
-
-    def accept_connection
-      @worker_pool.add_work @tcp_server.accept
-    end
+    def accept_client_connections
+      ready_inputs, _, _ = IO.select([@tcp_server, @signal_reader])
 
-    def process_signal
-      @signal.send @signal_reader.read_nonblock(1)
-    end
+      if ready_inputs.include?(@tcp_server)
+        @worker_pool.push @tcp_server.accept
+      end
 
-    def shutdown_worker_pool
-      logger.info "Shutting down worker pool"
-      @worker_pool.shutdown(@shutdown_timeout)
+      if ready_inputs.include?(@signal_reader)
+        @signal.send @signal_reader.read_nonblock(1)
+      end
     end
 
-    def clear_thread
-      @work_loop_thread = nil
+    def teardown
+      logger.info "Stopping work loop..."
+      unless @signal.halt?
+        logger.info "Shutting down worker pool"
+        @worker_pool.shutdown(@shutdown_timeout)
+      end
+      unless @signal.pause?
+        logger.info "Closing TCP server connection"
+        self.stop_listen
+      end
```

Comment:
@jcredding minor, and I just noticed b/c of this awesome set of changes, but don't you want to stop listening *before* shutting down the worker pool?